test/cpp/qps/client_async.cc

```diff
@@ -949,10 +949,125 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
       return nullptr;
   }
 }
+
 std::unique_ptr<Client> CreateGenericAsyncStreamingClient(
     const ClientConfig& args) {
   return std::unique_ptr<Client>(new GenericAsyncStreamingClient(args));
 }
 
+class CallbackClient
+    : public ClientImpl<BenchmarkService::Stub, SimpleRequest> {
+ public:
+  CallbackClient(const ClientConfig& config)
+      : ClientImpl<BenchmarkService::Stub, SimpleRequest>(
+            config, BenchmarkStubCreator) {
+    num_threads_ =
+        config.outstanding_rpcs_per_channel() * config.client_channels();
+    SetupLoadTest(config, num_threads_);
+    threads_done_ = 0;
+  }
+
+  virtual ~CallbackClient() {}
+
+  virtual bool InitThreadFuncImpl(size_t thread_idx) { return true; }
+
+  virtual bool ThreadFuncImpl(Thread* t, size_t thread_idx) = 0;
+
+  void ThreadFunc(size_t thread_idx, Thread* t) override {
```

Comment:
Ok, I see that you tried to pigeonhole this into ThreadFunc like the sync API benchmark, but could you instead make the code based on the async API benchmark, just without completion queues etc? Because as it stands, the ThreadFunc is just going to return quickly and then get reinvoked and you're going to have a bunch of active RPCs going on at once. So this won't be ping-pong at all in any case, but will be more like unbounded (unlike the existing models which bound the number of RPCs that are active at once)

Reply:
I see. Thanks Vijay. I guess I need to clarify my understanding of when a RPC is terminated while using Callback API. Will meet up with you discuss about it. Also will study async client more to address the issues raised in this review.